Sources/NIOHTTP2/HTTP2FrameParser.swift

```diff
@@ -1163,13 +1163,14 @@ fileprivate extension ByteBuffer {
             return nil
```

Comment:
I don't really like the pre-calculation up here ^^^ and the `!`s down there. I think this should be replaced with

```swift
let saveSelf = self
guard let lenHigh = self.readInteger(as: UInt16.self),
      let lenLow = self.readInteger(as: UInt8.self),
      let lenLow = self.readInteger(as: UInt8.self),
      let type = self.readInteger(as: UInt8.self),
      let flags = self.readInteger(as: UInt8.self),
      let streamID = self.readInteger(as: UInt32.self) else {
    self = saveSelf
    return nil
}

return FrameHeader(...)
```

Reply:
I disagree.

Specifically, the form you proposed does repeated and unnecessary bounds checking, incurring a bunch of potentially unnecessary pointer dereferences. Done maliciously it substantially increases the work needed to parse a frame header by forcing a quadratic behaviour when parsing frame headers.

Reply:
There are no pointer dereferences.

Reply:
Sure there are, reading data from the `ByteBuffer` is a pointer dereference.

Reply:
the bounds checks are integer comparisons. I agree that we should add a way that can hoist the bounds checks but we don't have one today. Going into unsafe-land is not a good excuse to reduce a constant factor

Reply:
For the future, I want something like

```
self.read( integer(UInt16.self) <*> integer(UInt8.self) <*> integer(UInt8.self) <*> integer(UInt8.self) )
```

that does the bounds check only once. But we're not there yet.